OmarAlFarajat/Always-Day
# Unity Project Cleanup and Restructuring

# What issue is this PR solving?
- Disorganization in the main scene hierarchy
- Disorganization in the the project folders and files structure
- Minor: Platform bug #37
- Minor: Music not looping and too loud
- Colliders for simple path shapes are using mesh colliders when they could use box colliders instead.

# Proposed Changes
- Main scene has been grouped into 4 zones using an empty game object as a parent: Start, Red, Green, Blue. The parent game objects have the format [COLOR/START]_ZONE
- Each of the 3 colored zones more or less has the following in their sub-hierarchy: 
   - Ghost
   - Obelisk
   - Block (used to block the path after defeating the ghost)
   - Hazards (things that shoot projectiles)
   - Restart (1)
   - Restart 2
   - Killzones
   - Path (the static path of the zone which excludes platform areas)
   - Platforms (for sections with moving platforms for red and blue only)
   - Scenery (for decoration, like trees and rocks)
   - Ocean (each zone happens to have a portion of visible ocean with its own dedicated water plane)
- The start zone has some like the above and its own sub-hierarchy:
   - RGB Obelisk
   - Barriers (to prevent player from walking out-of-bounds)
   - Main restart 
   - Ocean
   - Scenery
- Old scene files have been deleted and the important scenes consolidated to: Main Menu, Main Scene, End Scene
- In /Assets/Scenes/Personal/, I've made personal copies of the new Main Scene for each of you to work off of. 
- All "loose" files in the Assets folder and subfolders have been grouped into their own subfolders. I did this to the best of my abilities. If you have an issue with some of the folder structure, please let me know. 
- I deleted some files that simply weren't being used at all in any of the scenes (scripts, prefabs, materials, etc). Rest assured I double checked all references and tested the game.
- For each zone path, meshes that are box shaped now have box colliders instead of mesh colliders for cleaner movement by player and improved performance / less resource intensive. 
- Minor bugs with platforming and music not looping have been fixed. 

# Additional Info
When the game is played in the Unity editor, you will see numerous projectiles being spawned and de-spawned continuously. Besides being resource intensive, this also makes it hard to navigate the hierarchy for debugging. This will be solved in the future with the implementation of object pooling ( #44 )

# How to reproduce the changes
- Open the main scene file
- Observe any features you worked directly on and see if they behave properly after the restructure
- Observe the folders/files structure and see if anything important is missing
- Please comment @OmarAlFarajat here in this PR if you have any questions about your features in this restructure

# Main Scene Game Object Hierarchy (Before & After)

|      BEFORE     |       AFTER      |
|:--------------------:|:--------------------:|
| ![OldHierarchy](https://user-images.githubusercontent.com/40572852/97526010-9c130d00-197e-11eb-9904-244afe1067e1.png) | ![NewHierarchy](https://user-images.githubusercontent.com/40572852/97525950-74bc4000-197e-11eb-83ad-944447d5310d.png) |

# Sub-Hierarchies for Each Zone (Highlighted Selection)

|      **START_ZONE**     |
|:--------------------:|
| ![StartZone](https://user-images.githubusercontent.com/40572852/97526245-1ba0dc00-197f-11eb-98d1-5a5cc03ff75e.png) |

|      **RED_ZONE**     |
|:--------------------:|
| ![RedZone](https://user-images.githubusercontent.com/40572852/97526297-31ae9c80-197f-11eb-8e10-dec9769ed642.png) |

|     **GREEN_ZONE**     |
|:--------------------:|
| ![GreenZone](https://user-images.githubusercontent.com/40572852/97526310-3a06d780-197f-11eb-91f8-444dce35c4ca.png) |

|     **BLUE_ZONE**     |
|:--------------------:|
| ![BlueZone](https://user-images.githubusercontent.com/40572852/97526328-42f7a900-197f-11eb-9d76-97ae79441586.png) |
